src/test/HST/ApplicationTests.hs

```diff
@@ -0,0 +1,139 @@
+
+import           Language.Haskell.Exts          ( Module
+                                                , ParseResult(..)
+                                                , parseModule
+                                                )
+import           Language.Haskell.Exts.Pretty   ( Pretty
+                                                , prettyPrint
+                                                )
+
+import           HST.Application                ( processModule
+                                                , specialCons
+                                                )
+import           HST.Environment.FreshVars      ( PMState(..)
+                                                , evalPM
+                                                )
+
+-- | Test group for basic unit tests
+applicationTests :: Spec
+applicationTests = describe "Basic unit tests" tests
+
+-- | Parses a given string to a module and fails if parsing is not
+--   successful.
+parseTestModule :: String -> IO (Module ())
+parseTestModule modStr = case parseModule modStr of
+  ParseOk modul        -> return $ void modul
+  ParseFailed _ errMsg -> assertFailure errMsg
+
+-- | The default state used for pattern matching.
+defaultState :: PMState
+defaultState = PMState { nextId      = 0
+                       , constrMap   = specialCons
+                       , matchedPat  = []
+                       , trivialCC   = False
+                       , opt         = True
+                       , debugOutput = ""
+                       }
+
+-- | Test cases for 'processModule'.
+tests :: Spec
+tests = do
+  it "should accept a simple function" $ do
+    mod1 <- parseTestModule "module A where\nf :: a -> a\nf x = x"
+    let mod2 = evalPM (processModule mod1) defaultState
+    mod2 `prettyShouldBe` mod1
+  it "should transform pattern matching into case expressions" $ do
+    mod1 <- parseTestModule $ unlines
+      [ "module A where"
+      , "lengthL :: [a] -> Int"
+      , "lengthL [] = 0"
+      , "lengthL(_:xs) = 1 + lengthL xs"
+      ]
+    expected <- parseTestModule $ unlines
+      [ "module A where"
+      , "lengthL :: [a] -> Int"
+      , "lengthL a0 = case a0 of"
+      , "  []    -> 0"
+      , "  a1:a2 -> 1 + lengthL a2"
+      ]
+    let mod2 = evalPM (processModule mod1) defaultState
+    expected `prettyShouldBe` mod2
+  it "should transform pattern matching in a partial function" $ do
+    mod1 <- parseTestModule
+      $ unlines ["module A where", "head :: [a] -> a", "head (x:xs) = x"]
+    let mod2 = evalPM (processModule mod1) defaultState
+    expected <- parseTestModule $ unlines
+      [ "module A where"
+      , "head :: [a] -> a"
+      , "head a0 = case a0 of"
+      , "  a1 : a2 -> a1"
+      , "  [] -> undefined"
+      ]
+    mod2 `prettyShouldBe` expected
+  it "should accept a simple guarded expression" $ do
+    mod1 <-
+      parseTestModule
+      $  "module A where\n"
+      ++ "id :: a -> a\n"
+      ++ "id x | otherwise = x"
```

Comment:
There are still some test cases that don't use `unlines`.

```suggestion
    mod1 <- parseTestModule $ unlines
      [ "module A where"
      , "id :: a -> a"
      , "id x | otherwise = x"
      ]
```